Add EIP: Zero-Knowledge AI-Generated Content Token (#7007)
ethereum/EIPs: EIPS/eip-7007.md

---
eip: 7007

> title: zkML AIGC-NFTs

Reviewer: This title is still a bit too much alphabet soup. How about:

```suggestion
title: Zero-Knowledge AI Generated Content Token
```

description: An ERC-721 extension interface for zkML based AIGC-NFTs.

> author: Cathie So (@socathie), Xiaohang Yu <xhyumiracle@hyperoracle.io>, Huaizhe Xu (@HuaizheXu), Kartin <kartin@hyperoracle.io>

Reviewer: You'll want to add your GitHub username as an author or else you won't be able to edit the proposal without getting approvals from the other authors.

discussions-to: https://ethereum-magicians.org/t/eip-7007-zkml-aigc-nfts-an-erc-721-extension-interface-for-zkml-based-aigc-nfts/14216
status: Draft
type: Standards Track
category: ERC
created: 2023-05-10
requires: 165, 721
---

## Abstract

> The zkML AIGC-NFTs standard is an extension of the [ERC-721](./eip-721.md) token standard for AI Generated Content. It proposes a set of interfaces for basic interactions and enumerable interactions for AIGC-NFTs. The standard includes a new mint event and a JSON schema for AIGC-NFT metadata. Additionally, it incorporates zkML capabilities to enable verification of AIGC-NFT ownership. In this standard, the `tokenId` is indexed by the `prompt`.

Reviewer: Please expand abbreviations the first time they are used (eg. `The zebra kebab making license (zkML)...`)

> The Zero-Knowledge Machine Lerning (zkML) AI Generated Content (AIGC) non-fungible token (NFT) standard is an extension of the [ERC-721](./eip-721.md) token standard for AIGC. It proposes a set of interfaces for basic interactions and enumerable interactions for AIGC-NFTs. The standard includes a new mint event and a JSON schema for AIGC-NFT metadata. Additionally, it incorporates zkML capabilities to enable verification of AIGC-NFT ownership. In this standard, the `tokenId` is indexed by the `prompt`.

Reviewer: ```suggestion
The Zero-Knowledge Machine Learning (zkML) AI Generated Content (AIGC) non-fungible token (NFT) standard is an extension of the [ERC-721](./eip-721.md) token standard for AIGC. It proposes a set of interfaces for basic interactions and enumerable interactions for AIGC-NFTs. The standard includes a new mint event and a JSON schema for AIGC-NFT metadata. Additionally, it incorporates zkML capabilities to enable verification of AIGC-NFT ownership. In this standard, the `tokenId` is indexed by the `prompt`.
```

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

**Every compliant contract must implement the [ERC-7007](./eip-7007.md), [ERC-721](./eip-721.md), and [ERC-165](./eip-165.md) interfaces.**

The zkML AIGC-NFTs standard includes the following interfaces: 

> ERC-7007: Defines a mint event and a mint function for minting AIGC-NFTs. It also includes a verify function to check the validity of a combination of prompt and proof using zkML techniques.

Reviewer: I'm assuming the bot made you put a dash here? If you use backticks to indicate inline code, the bot won't complain:

```suggestion
`IERC7007`: Defines a mint event and a mint function for minting AIGC-NFTs. It also includes a verify function to check the validity of a combination of prompt and proof using zkML techniques.
```

```solidity
pragma solidity ^0.8.18;

/**
 * @dev Required interface of an ERC7007 compliant contract.
 * Note: the ERC-165 identifier for this interface is 0x7e52e423.
 */
interface IERC7007 is IERC165, IERC721 {
    /**
     * @dev Emitted when `tokenId` token is minted.
     */
    event Mint(
        uint256 indexed tokenId,
        bytes indexed prompt,
        bytes indexed aigcData,
        string uri,
        bytes proof
    );

    /**
     * @dev Mint token at `tokenId` given `prompt`, `aigcData`, `uri` and `proof`.
     *
     * Requirements:
     * - `tokenId` must not exist.'
     * - verify(`prompt`, `aigcData`, `proof`) must return true.
     *
     * Optional:
     * - `proof` should not include `aigcData` to save gas.
     */
    function mint(
        bytes calldata prompt,
        bytes calldata aigcData,
        string calldata uri,
        bytes calldata proof

>     ) external returns (uint256 tokenId);

Reviewer: Generally minting is an application-specific operation. In other words, it is unusual to standardize `mint` because how it works is usually tied pretty closely to each token.

The question I'll need you to answer is: will anyone other than the deployer of the contract (or the deployer's specific dapp, like OpenSea) be calling the mint function?

If not, you don't really need to standardize it, though I think adding the ML metadata extensions make sense.

Reviewer 2: @SamWilsn This seems a peer review rather than an editoral review. Maybe put it in the FEM?

        },

        "prompt": {
            "type": "string",
            "description": "Identifies the prompt from which this AIGC NFT generated"
        },
        "aigc_type": {
            "type": "string",
            "description": "image/video/audio..."
        },
        "aigc_data": {
            "type": "string",
            "description": "A URI pointing to a resource with mime type image/* representing the asset to which this AIGC NFT represents."
        }
    }
}
```

## Rationale

The zkML AIGC-NFTs standard aims to extend the existing [ERC-721](./eip-721.md) token standard to accommodate the unique requirements of AI Generated Content NFTs representing models in a collection. This standard provides interfaces to use zkML to verify whether or not the AIGC data for an NFT is generated from a certain ML model with certain input (prompt). The proposed interfaces allow for additional functionality related to minting, verifying, and enumerating AIGC-NFTs. Additionally, the metadata schema provides a structured format for storing information related to AIGC-NFTs, such as the prompt used to generate the content and the proof of ownership.

> With this standard, model owners can publish their trained model and its ZKP verifier to Ethereum. Any user can claim an input (prompt) and publish the inference task, any node that maintains the model and the proving circuit can perform the inference and proving, then submit the output of inference and the ZK proof for the inference trace into the verifier that is deployed by the model owner. The user that initiates the inference task will own the output for the inference of that model and input (prompt).

Reviewer: Both of these paragraphs seem more like they belong in the motivation section. The rationale section should explain individual technical decisions made within the EIP itself, while the motivation section should justify the EIP as a whole.
